### PR TITLE
Transform code flows

### PR DESCRIPTION
**Pull request by EasyRhin0**
Process result.relatedLocations
Remove result.snippet

### CODE REVIEW

**Comment by EasyRhin0** [comment]
Notes:
I'm putting v1 AnnotatedCodeLocationKind in the new codeFlowLocation.kind property
I'm stashing the "essential" property even though it's obsolete

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 1]
```diff
@@ -1453,5 +1483,336 @@ public void SarifTransformerTests_ToCurrent_TwoResultsWithFixes()
...
+            string v1LogText =
+@"{
+  ""version"": ""1.0.0"",
+  ""runs"": [
+    {
+      ""tool"": {
+        ""name"": ""CodeScanner"",
+        ""semanticVersion"": ""2.1.0""
+      },
+      ""logicalLocations"": {
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: I don't think we need v1 `logicalLocations` to test v1 => v2 code flows.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 2]
```diff
@@ -1453,5 +1483,336 @@ public void SarifTransformerTests_ToCurrent_TwoResultsWithFixes()
...
+        },
+        ""collections"": {
+          ""name"": ""collections"",
+          ""kind"": ""namespace""
+        }
+      },
+      ""results"": [
+        {
+          ""ruleId"": ""C2001"",
+          ""formattedRuleMessage"": {
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: We don't need rule metadata to test code flows. Just use `result.message`.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 1 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: The locations have fullyQualifiedLogicalNames

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 2 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: OK

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 3]
```diff
@@ -1453,5 +1483,336 @@ public void SarifTransformerTests_ToCurrent_TwoResultsWithFixes()
...
+                              ""text"": ""int *ptr;""
+                            }
+                          }
+                        },
+                        ""fullyQualifiedLogicalName"": ""collections::list:add"",
+                        ""message"": {
+                          ""text"": ""Variable \""ptr\"" declared.""
+                        }
+                      },
+                      ""kind"": ""Declaration"",
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Even though the v2 `codeFlowLocation` has a property `kind`, it doesn't mean the same thing as the v1 `annotatedCodeLocation.kind`. I can explain if you're interested. Put this in the property bag.

Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: OK

File: src/Sarif/Visitors/SarifTransformerUtilities.cs [hunk 4]
```diff
@@ -192,5 +192,17 @@ public static BaselineState CreateBaselineState(BaselineStateVersionOne v1Baseli
...
+        public static CodeFlowLocationImportance CreateCodeFlowLocationImportance(AnnotatedCodeLocationImportanceVersionOne v1AnnotatedCodeLocationImportance)
+        {
+            switch (v1AnnotatedCodeLocationImportance)
+            {
+                case AnnotatedCodeLocationImportanceVersionOne.Essential:
+                    return CodeFlowLocationImportance.Essential;
+                case AnnotatedCodeLocationImportanceVersionOne.Important:
+                    return CodeFlowLocationImportance.Important;
+                default:
+                    return CodeFlowLocationImportance.Unimportant;
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifTransformerUtilities.cs]: The default for v1's `annotatedCodeLocation.importance` property was `important`.

(The same is true in v2, but that's neither here nor there, since you're processing a v1 value here.)

Comment (EasyRhin0) [reply src/Sarif/Visitors/SarifTransformerUtilities.cs]: Fixed

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs [hunk 5]
```diff
@@ -33,6 +35,85 @@ public override SarifLogVersionOne VisitSarifLogVersionOne(SarifLogVersionOne v1
...
+                };
+
+                if (v1CodeFlow.Locations != null && v1CodeFlow.Locations.Count > 0)
+                {
+                    _codeFlowLocationNestingLevel = 0;
+
+                    if (v1CodeFlow.Locations[0].Step == 0)
+                    {
+                        // If the steps are zero-based, add 1 to comply with the v2 spec
+                        _codeFlowLocationStepAdjustment = 1;
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Nice! 😄 

But (I hate to be a killjoy 😄 ): Bug: This means that `step` will not round-trip. If it was 0-based on the way in from v1 to v2, you'll have forgotten about that when you convert back to v1. You need to put the v1 `step` in the property  bag.

... as usual, unless and until we get tired of worrying about this stuff and stick the whole v1 run in the v2 run's property bag.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 1 above)
Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Sure, but why does v1 `logicalLocations` have to exist? You'll create v2 `logicalLocations` when you notice the FQLNs in the code flow.

Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: The spec says that _if_ `run.logicalLocations` exists, _then_ `location.FQLN` must match one of its property names. But the spec does not say that `run.logicalLocations` has to exist.

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs [hunk 6]
```diff
@@ -33,6 +35,85 @@ public override SarifLogVersionOne VisitSarifLogVersionOne(SarifLogVersionOne v1
...
+                };
+
+                if (v1AnnotatedCodeLocation.Kind == AnnotatedCodeLocationKindVersionOne.Call)
+                {
+                    _codeFlowLocationNestingLevel++;
+                }
+                else if (v1AnnotatedCodeLocation.Kind == AnnotatedCodeLocationKindVersionOne.CallReturn)
+                {
+                    _codeFlowLocationNestingLevel--;
+                }
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Good.

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs [hunk 7]
```diff
@@ -263,6 +344,40 @@ internal Location CreateLocation(LocationVersionOne v1Location)
...
+        internal Location CreateLocation(AnnotatedCodeLocationVersionOne v1AnnotatedCodeLocation)
+        {
+            Location location = null;
+
+            if (v1AnnotatedCodeLocation != null)
+            {
+                location = new Location
+                {
+                    Annotations = v1AnnotatedCodeLocation.Annotations?.SelectMany(a => a.Locations, (a, pl) => CreateRegion(pl)).ToList(),
+                    FullyQualifiedLogicalName = v1AnnotatedCodeLocation.LogicalLocationKey ?? v1AnnotatedCodeLocation.FullyQualifiedLogicalName,
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Good.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 8]
```diff
@@ -1453,5 +1483,336 @@ public void SarifTransformerTests_ToCurrent_TwoResultsWithFixes()
...
+                      ""module"": ""platform"",
+                      ""importance"": ""essential"",
+                      ""properties"": {
+                        ""sarifv1/annotations"": null,
+                        ""sarifv1/essential"": false,
+                        ""sarifv1/fullyQualifiedLogicalName"": ""collections::list:add"",
+                        ""sarifv1/id"": 0,
+                        ""sarifv1/logicalLocationKey"": null,
+                        ""sarifv1/taintKind"": 0,
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Bug: This won't round-trip. v1 `taintKind` was a string with enumerated values `"source"`, `"sink"`, `"sanitizer"`.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 9]
```diff
@@ -1453,5 +1483,336 @@ public void SarifTransformerTests_ToCurrent_TwoResultsWithFixes()
...
+                        ""sarifv1/target"": null,
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: We need to test that this works for non-null values of `target`, `targetKey`, and `values`.

Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: ... unless and until we put the whole run in the property bag! I'm going to ping Michael on this.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 8 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Fixed

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 10]
```diff
@@ -1453,5 +1483,336 @@ public void SarifTransformerTests_ToCurrent_TwoResultsWithFixes()
...
+                        ""sarifv1/targetKey"": null,
+                        ""sarifv1/threadId"": 52,
+                        ""sarifv1/values"": null
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Bug: You need to bag v1  `annotatedCodeLocation.state` and `targetLocation` as well.

Until Michael lets us off the hook 😄.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 1 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: I'm sorry, I'm sure we went through this once before, but doesn't the presence of a FQLN imply a corresponding v1 logicalLocation?

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs [hunk 11]
```diff
@@ -263,6 +344,40 @@ internal Location CreateLocation(LocationVersionOne v1Location)
...
 
+        internal Location CreateLocation(AnnotatedCodeLocationVersionOne v1AnnotatedCodeLocation)
+        {
+            Location location = null;
+
+            if (v1AnnotatedCodeLocation != null)
+            {
+                location = new Location
+                {
+                    Annotations = v1AnnotatedCodeLocation.Annotations?.SelectMany(a => a.Locations, (a, pl) => CreateRegion(pl)).ToList(),
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Oh, wow, very nice. I had to look it up to remind myself that v1 `location.annotations` is of type `region[]`.

But there's a bug. The v1 `annotation`  object contained a message, and a set of locations to which the message applies. It was a terrible design for reasons I can go into offline if you like.

The point is, in the new design -- where the `region` object has a `message` property -- that v1 `annotatedCodeLocation.message` needs to go on each of the v2 `region`s you create. As the code stands, the message is lost.

We need to enhance the tests to detect this (by adding a couple of v1 `annotations` to one of the v1 `annotatedCodeLocation`s.

Comment (EasyRhin0) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Gotcha

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 1 above)
Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: No. If you want to tell the consumer more about the logical location whose FQLN is `location.FQLN`, you can create the `run.logicalLocations` dictionary,  add a property whose name is the value of `FQLN`, and set the property value to a `logicalLocation` object with the additional information. But you didn't have to do it in v1, and you still don't have to do it in v2.

Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: (Don't worry about asking the same question multiple times. I'm notorious for doing that -- just ask Michael 😄.)

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 3 above)
Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: ... or, don't bother doing that now, since we're not going to do this piecemeal bagging any more.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 10 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Done

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 9 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Done

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 10 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: State transfers to v2

Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: No v1 targetLocation property

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs (hunk 11 above)
Comment (EasyRhin0) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Done

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 10 above)
Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Ah right. We need a test for that.

Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: The "ah right, we need a test" was about `state`.

As to `targetLocation`, [are you sure](https://rawgit.com/sarif-standard/sarif-spec/master/Static%20Analysis%20Results%20Interchange%20Format%20(SARIF).html#annotatedcodelocation-target)?

Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: [Yes](https://github.com/oasis-tcs/sarif-spec/raw/master/Documents/WorkingDrafts/sarif-v1.0-wd01.docx)
[Schema](https://github.com/Microsoft/sarif-sdk/blob/sarif-v2/src/Sarif/Schemata/Sarif-1.0.0.schema.json)

Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Well, ok, that was a bug in the v1 schema. It didn't implement the v1 spec.

Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Done

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 1 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Clear, thanks

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 12]
```diff
@@ -1453,5 +1483,337 @@ public void SarifTransformerTests_ToCurrent_TwoResultsWithFixes()
...
+                      ""module"": ""platform"",
+                      ""importance"": ""essential"",
+                      ""properties"": {
+                        ""sarifv1/annotations"": null,
+                        ""sarifv1/essential"": false,
+                        ""sarifv1/fullyQualifiedLogicalName"": ""collections::list::add"",
+                        ""sarifv1/id"": 0,
+                        ""sarifv1/kind"": ""Declaration"",
+                        ""sarifv1/logicalLocationKey"": null,
+                        ""sarifv1/taintKind"": ""Sink"",
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Bug: The value in the v1 file was `"sink"`.

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs [hunk 13]
```diff
@@ -263,6 +346,43 @@ internal Location CreateLocation(LocationVersionOne v1Location)
...
 
+        internal Location CreateLocation(AnnotatedCodeLocationVersionOne v1AnnotatedCodeLocation)
+        {
+            Location location = null;
+
+            if (v1AnnotatedCodeLocation != null)
+            {
+                location = new Location
+                {
+                    Annotations = v1AnnotatedCodeLocation.Annotations?.SelectMany(a => a.Locations, (a, pl) => CreateRegion(pl, a.Message)).ToList(),
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: There is a problem here caused by the poor design of the v1 `annotation` object. You have a `location` object that refers to a file, say _file1.c_. The v1 `annotation` object contained an array of `physicalLocation` objects -- which meant that an annotation on _file1.c_  could refer to a location in _file2.c_. That makes no sense. That's why we changed the design and got rid of the `annotation` object. In v2, annotations can only refer to regions in the location's own file.

So what do you do if you encounter a v1 file that makes use of this senseless feature (as your test example does)? The best you can do is
1. Copy the whole set of v1 annotations to the property bag. You're doing that.
2. Populate the v2 `location.annotations` property with _only those annotations that refer to the file specified by_ `location.physicalLocation`.

You'll never encounter this; AFAIK no existing v1 producer uses annotations at all, still less in this senseless manner. But the code may as well be prepared for it.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 14]
```diff
@@ -1453,5 +1483,337 @@ public void SarifTransformerTests_ToCurrent_TwoResultsWithFixes()
...
+  ""version"": ""1.0.0"",
+  ""runs"": [
+    {
+      ""tool"": {
+        ""name"": ""CodeScanner"",
+        ""semanticVersion"": ""2.1.0""
+      },
+      ""results"": [
+        {
+          ""message"": ""C2001"",
```
[the remark this answers is not in the document]
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Fixed

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 12 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Changed

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs (hunk 13 above)
Comment (EasyRhin0) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Done
